File: scripts/strip-source-footer.py
#!/usr/bin/env python3
import urllib.request, urllib.parse, json, sys

TOKEN = "skbUvbYYIvf0Uwc43kqoHa7MX556BIABP7tNDQjW06yeBHY9ImiPeEjgMs87ZxlUafA5XRt6LXwn8d5Y9JcmDaZN13fvjxt6Tm3QgSAE8LqSvP6oU7zgF3W4dGb3jnjVIuBnZTICBsln2LHqgKjFIAybBohK6JCJWR8qHmP6CMhPVpsiPB79"
PROJECT = "vbnsqnkg"
BASE = f"https://{PROJECT}.api.sanity.io/v2024-01-01/data"

def sanity_query(q):
    url = f"{BASE}/query/production?query={urllib.parse.quote(q)}"
    req = urllib.request.Request(url, headers={"Authorization": f"Bearer {TOKEN}"})
    resp = urllib.request.urlopen(req)
    return json.loads(resp.read())["result"]

def sanity_mutate(mutations):
    url = f"{BASE}/mutate/production?returnIds=false"
    body = json.dumps({"mutations": mutations}).encode()
    req = urllib.request.Request(url, data=body, method="POST",
        headers={"Authorization": f"Bearer {TOKEN}", "Content-Type": "application/json"})
    resp = urllib.request.urlopen(req)
    return json.loads(resp.read())

import re

PATTERNS = [
    re.compile(r"<p[^>]*>[^<]*[Ss]ource:[^<]*[Vv]isit the original[^<]*</p>", re.IGNORECASE),
    re.compile(r"<p[^>]*>[^<]*[Ss]ource:[^<]*[Bb]earing Arms[^<]*</p>", re.IGNORECASE),
    re.compile(r"<p[^>]*>\s*<em>[^<]*[Ss]ource:[^<]*</em>\s*</p>", re.IGNORECASE),
    re.compile(r"\n?<p[^>]*>[^<]*visit the original[^<]*article[^<]*</p>", re.IGNORECASE),
    re.compile(r"\n?Source:[^\n]*visit the original[^\n]*", re.IGNORECASE),
    re.compile(r"\n?Source:[^\n]*[Bb]earing Arms[^\n]*", re.IGNORECASE),
]

def strip(body):
    if not body: return body, False
    out = body
    for p in PATTERNS:
        out = p.sub("", out)
    out = out.strip()
    return out, out != body

print("Fetching articles from Sanity...")
docs = sanity_query('*[_type == "newsArticle" && defined(body) && body != ""]{ _id, body }')
print(f"Total articles: {len(docs)}")

# Show sample tails
for i in range(min(3, len(docs))):
    tail = docs[i]["body"][-300:].replace("\n", " ")
    print(f"SAMPLE {i}: {tail}")

patched = 0
skipped = 0
BATCH = 50

for i in range(0, len(docs), BATCH):
    batch = docs[i:i+BATCH]
    mutations = []
    for doc in batch:
        out, changed = strip(doc["body"])
        if changed:
            mutations.append({"patch": {"id": doc["_id"], "set": {"body": out}}})
            patched += 1
        else:
            skipped += 1
    if mutations:
        sanity_mutate(mutations)
        print(f"Committed batch {i}-{i+BATCH} ({len(mutations)} patches)")

print(f"DONE. Patched: {patched} | Skipped: {skipped} | Total: {len(docs)}")

File: scripts/strip-source-footer.mjs
#!/usr/bin/env node
const TOKEN = process.env.SANITY_TOKEN
const PROJECT = 'vbnsqnkg'
const BASE = `https://${PROJECT}.api.sanity.io/v2024-01-01/data`

async function query(q) {
  const url = `${BASE}/query/production?query=${encodeURIComponent(q)}`
  const r = await fetch(url, { headers: { Authorization: `Bearer ${TOKEN}` } })
  if (!r.ok) throw new Error(`GROQ ${r.status}: ${await r.text()}`)
  return (await r.json()).result
}

async function mutate(mutations) {
  const r = await fetch(`${BASE}/mutate/production?returnIds=false`, {
    method: 'POST',
    headers: { Authorization: `Bearer ${TOKEN}`, 'Content-Type': 'application/json' },
    body: JSON.stringify({ mutations })
  })
  if (!r.ok) throw new Error(`Mutate ${r.status}: ${await r.text()}`)
  return r.json()
}

function strip(body) {
  if (!body) return { out: body, changed: false }
  let out = body
  out = out.replace(/<p[^>]*>[^<]*[Ss]ource:[^<]*[Vv]isit the original[^<]*<\/p>/gi, '')
  out = out.replace(/<p[^>]*>[^<]*[Ss]ource:[^<]*[Bb]earing Arms[^<]*<\/p>/gi, '')
  out = out.replace(/<p[^>]*>\s*<em>[^<]*[Ss]ource:[^<]*<\/em>\s*<\/p>/gi, '')
  out = out.replace(/\n?<p[^>]*>[^<]*visit the original[^<]*article[^<]*<\/p>/gi, '')
  out = out.replace(/\n?Source:[^\n]*visit the original[^\n]*/gi, '')
  out = out.replace(/\n?Source:[^\n]*[Bb]earing Arms[^\n]*/gi, '')
  out = out.trim()
  return { out, changed: out !== body }
}

const docs = await query('*[_type == "newsArticle" && defined(body) && body != ""]{ _id, body }')
console.log('Total articles:', docs.length)

// Print 3 sample tails
for (let i = 0; i < Math.min(3, docs.length); i++) {
  console.log('SAMPLE', i, docs[i].body.slice(-250).replace(/\n/g,' '))
}

let patched = 0, skipped = 0
const BATCH = 50
for (let i = 0; i < docs.length; i += BATCH) {
  const batch = docs.slice(i, i + BATCH)
  const mutations = []
  for (const doc of batch) {
    const { out, changed } = strip(doc.body)
    if (changed) {
      mutations.push({ patch: { id: doc._id, set: { body: out } } })
      patched++
    } else skipped++
  }
  if (mutations.length) {
    await mutate(mutations)
    console.log('Committed batch', i, '-', i + BATCH, '(', mutations.length, 'patches)')
  }
}
console.log('DONE. Patched:', patched, '| Skipped:', skipped, '| Total:', docs.length)
